streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -635,7 +635,9 @@ private void maybeAbortTransactionAndCloseRecordCollector(final boolean isZombie
 
                 // can be ignored: transaction got already aborted by brokers/transactional-coordinator if this happens
             }
+        }
 
+        if (eosEnabled) {
```

Comment:
We actually did not call `close()` for this case on purpose. IIRC, the producer contract is to not call _any_ method (not even `close()`) after a `ProducerFencedException` was thrown by the producer.

\cc @hachikuji to confirm.

Reply:
This change also does not seem to align to what the JIRA describes. In fact, it seems that the old code ignores the `isZombie` flag and calls `close()` if `eosEnable=true` (what seems wrong). The fix seems to ensure that we call `close()` only for the non-fencing case though.

Reply:
>it seems that the old code ignores the `isZombie` flag and calls close() if eosEnable=true

Do you mean the new code? The previous code only closed the `recordCollector` (which in turn closes the `producer`) when `isZombie=false`.  This was leading to streams leaking producers that never got closed after a rebalance from a fenced producer.

Reply:
I was confused initially -- sorry for the confusion -- deleted my other comment already...

Reply:
@mjsax @bbejeck Just to clarify, I think we should always close the producer even if it was fenced. Resources like network connections only get cleaned up in `KafkaProducer.close()`.

Reply:
Thanks for confirming @hachikuji!